RotateOnBan LP + rotateOnBan config

ByteAlex opened the pull request:
Title says it all, I wasn't able to fully test it locally, cause my IPv6 subnet wasn't setup properly, but maybe you can give it a short test (or I'll do as soon as I get a working IPv6 subnet)

Frederikam commented on LavalinkServer/application.yml.example:
```diff
@@ -18,14 +18,11 @@ lavalink:
-    # List of IPs in the block to exclude. Useful if your gateway is a member of the same block.
-    #excludedIps: ["...", "..."]
+    ratelimit:
+      #ipVersion: '6'
```
It would be nice if Lavalink would just detect which family

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/RoutePlannerService.kt:
```diff
@@ -0,0 +1,44 @@
+import java.util.function.Predicate
+
+@Service
+class RoutePlannerService(rateLimitConfig: RateLimitConfig) {
```
This should probably be merged with `AudioPlayerConfiguration`

The route planner is practically a part of the player config.

Frederikam replied:
`RateLimitConfig` can probably be null. Should be `RateLimitConfig?` unless you want old configs to break.

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/RoutePlannerService.kt:
```diff
@@ -0,0 +1,44 @@
+class RoutePlannerService(rateLimitConfig: RateLimitConfig) {
+
+
+    private final val routePlanner: AbstractRoutePlanner?
```
This is redundant. Just use one bean function. The bean will just be run once.

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/RoutePlannerService.kt:
```diff
@@ -0,0 +1,44 @@
+                else -> throw RuntimeException("Invalid IP Block, make sure to provide a valid CIDR notation")
+            }
+            when {
+                rateLimitConfig.strategy == "RotateOnBan" -> RotatingIpRoutePlanner(ipBlock, filter)
```
Trim and ignore this string for good measure.

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/RoutePlannerService.kt:
```diff
@@ -0,0 +1,44 @@
+    private final val routePlanner: AbstractRoutePlanner?
+
+    init {
+        routePlanner = if(rateLimitConfig.ipBlock.isNotEmpty()) {
```
You could add `if(rateLimitConfig.ipBlock.isEmpty()) return null` (in a function) and save the long if block.

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/player/RoutePlannerRestHandler.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Base64;
+
+@RestController
+public final class RoutePlannerRestHandler {
```
These routes should be documented

Frederikam replied:
Also, is this endpoint subject to authentication?

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/player/RoutePlannerRestHandler.java:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        try {
+            final InetAddress address = InetAddress.getByAddress(Base64.getDecoder().decode(ip));
```
Why base64?

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/player/RoutePlannerRestHandler.java:
```diff
@@ -0,0 +1,46 @@
+
+    @GetMapping(value = "/routeplanner/freeIp", produces = "application/json")
+    @ResponseBody
+    public ResponseEntity<String> freeSingleIp(HttpServletRequest request, @RequestParam String ip) {
```
This should be a POST request. Don't use a request parameter, use a body.

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/player/RoutePlannerRestHandler.java:
```diff
@@ -0,0 +1,46 @@
+    @ResponseBody
+    public ResponseEntity<String> freeSingleIp(HttpServletRequest request, @RequestParam String ip) {
+
+        final AbstractRoutePlanner activePlanner = AbstractRoutePlanner.getActivePlanner();
```
This looks like static abuse. Could you not just get the planner from dependency injection?

itslukej commented:
I think it would be important to note in the readme that Linux AnyIP should be configured on the CIDR block for correct operation.

https://blog.widodh.nl/2016/04/anyip-bind-a-whole-subnet-to-your-linux-machine/
https://blog.cloudflare.com/how-we-built-spectrum/#assigningmillionsofipstoaserver

ByteAlex commented:
Static abuse in Lavaplayer was removed, yet sedmelluq will change the implementation a little bit, but as far as I understood he will merge my PR

ByteAlex commented:
Uses official LP release now with latest versions @Frederikam

ByteAlex commented:
* Added ratelimit config section
* When ratelimit config is set up, a AbstractRoutePlanner is initialized from it (according to the strategy) and the YoutubeRotator extension from lavaplayer will be set up with it
* Added RoutePlannerRestHandler (For monitoring RoutePlanner statistics / bans)

Frederikam commented on LavalinkServer/application.yml.example:
```diff
@@ -18,14 +18,11 @@ lavalink:
-    # List of IPs in the block to exclude. Useful if your gateway is a member of the same block.
-    #excludedIps: ["...", "..."]
+    ratelimit:
+      #ipBlocks: ["127.0.0.1/31", "127.0.0.3/32"]
```
These are poor examples

Frederikam commented on LavalinkServer/application.yml.example:
```diff
@@ -18,14 +18,11 @@ lavalink:
+    ratelimit:
+      #ipBlocks: ["127.0.0.1/31", "127.0.0.3/32"]
+      #excludedIps: ["...", "..."]
+      #strategy: 'RotateOnBan' # RotateOnBan | LoadBalance | NanoSwitch | RotatingNanoSwitch
```
Use double quotes here for consistency

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/AudioPlayerConfiguration.kt:
```diff
@@ -10,45 +10,40 @@ import com.sedmelluq.discord.lavaplayer.source.soundcloud.SoundCloudAudioSourceM
-                YoutubeAudioSourceManager(serverConfig.isYoutubeSearchEnabled)
+            val youtube = YoutubeAudioSourceManager(serverConfig.isYoutubeSearchEnabled)
+            if (routePlanner != null) {
+                YoutubeIpRotator.setup(youtube, routePlanner)
```
So we're back to static abuse now? Even if this is lavaplayer it still concerns me.

ByteAlex replied:
It's based on sed's decision. He implemented it like that

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/AudioPlayerConfiguration.kt:
```diff
@@ -69,4 +64,42 @@ class AudioPlayerConfiguration {
+            log.debug("No rate limit config block found, skipping setup of route planner")
+            return null
+        }
+        val ipBlockList = ArrayList(rateLimitConfig.ipBlocks)
```
Avoid the use of `ArrayList` in Kotlin. Convert to a mutable list instead.

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/AudioPlayerConfiguration.kt:
```diff
@@ -69,4 +64,42 @@ class AudioPlayerConfiguration {
+        }
+        val ipBlockList = ArrayList(rateLimitConfig.ipBlocks)
+        if (rateLimitConfig.ipBlock.isNotEmpty()) {
+            log.warn("Usage of deprecated `ipBlock` found, please use `ipBlocks` list instead!")
```
Don't add legacy support for something that was not formally released.

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/AudioPlayerConfiguration.kt:
```diff
@@ -69,4 +64,42 @@ class AudioPlayerConfiguration {
+            ipBlockList.add(rateLimitConfig.ipBlock)
+        }
+        if (ipBlockList.isEmpty()) {
+            log.debug("List of ip blocks is empty, skipping setup of route planner")
```
Consider using info logging

Frederikam commented on LavalinkServer/application.yml.example:
```diff
@@ -18,14 +18,11 @@ lavalink:
-
-    # List of IPs in the block to exclude. Useful if your gateway is a member of the same block.
-    #excludedIps: ["...", "..."]
+    ratelimit:
```
Should probably also be commented out

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/config/ServerConfig.kt:
```diff
@@ -36,6 +36,4 @@ class ServerConfig {
     var isYoutubeSearchEnabled = true
     var isSoundcloudSearchEnabled = true
-    var balancingBlock = ""
-    var excludedIps: List<String> = emptyList()
```
It would make sense to add `RateLimitConfig` here

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+    fun getStatus(request: HttpServletRequest): ResponseEntity<String> {
+        val jsonObject = JSONObject()
+        if (routePlanner == null) {
+            jsonObject.put("status", "E")
```
Use HTTP status codes instead of this

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+        jsonObject.put("status", "OK")
+        jsonObject.put("class", routePlanner?.javaClass?.canonicalName ?: "Error")
+        jsonObject.put("details", getDetailBlock(routePlanner))
+        return ResponseEntity.ok(jsonObject.toString())
```
I believe `#toString()` is redundant

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+        return try {
+            val body = JSONObject(requestBody)
+            val address = InetAddress.getByName(body.getString("address"))
+            routePlanner?.freeAddress(address)
```
Use `!!` not `?.`

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+            jsonObject.put("status", "E")
+            jsonObject.put("message", "Invalid request body: " + exception.message)
+            ResponseEntity.badRequest().body(jsonObject.toString())
+        } catch (exception: UnknownHostException) {
```
Just let Spring handle the `UnknownHostException`

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+     * Removes a single address from the addresses which are currently marked as failing
+     */
+    @PostMapping(value = ["/routeplanner/free/address"], consumes = ["application/json"], produces = ["application/json"])
+    fun freeSingleAddress(request: HttpServletRequest, @RequestBody requestBody: String): ResponseEntity<String> {
```
Request body should be a POJO

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+     * Removes all addresses from the list which holds the addresses which are marked failing
+     */
+    @PostMapping(value = ["/routeplanner/free/all"], consumes = ["application/json"], produces = ["application/json"])
+    fun freeAllAddresses(request: HttpServletRequest, @RequestBody requestBody: String): ResponseEntity<String> {
```
`requestBody` is unused

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+        return try {
+            routePlanner?.freeAllAddresses()
+            jsonObject.put("status", "OK")
+            jsonObject.put("message", "address successfully unmarked")
```
Inconsistent capitalization

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+            return ResponseEntity.badRequest().body(jsonObject.toString())
+        }
+        return try {
+            routePlanner?.freeAllAddresses()
```
Don't use `?.`

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+import javax.servlet.http.HttpServletRequest
+
+@RestController
+class RoutePlannerRestHandler(val routePlanner: AbstractRoutePlanner?) {
```
Consider replacing all JSON with POJOs. Let Spring deal with all the (de)serialization.

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+    private fun getDetailBlock(planner: AbstractRoutePlanner?): JSONObject {
+        val jsonObject = JSONObject()
+
+        if (planner is RotatingIpRoutePlanner) {
```
Replace if statements with `when`

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        if(planner is NanoIpRoutePlanner) {
+            jsonObject.put("estAddressIndex", planner.currentAddress)
```
No idea what `estAddressIndex` means

Frederikam commented on LavalinkServer/src/main/java/lavalink/server/routeplanner/RoutePlannerRestHandler.kt:
```diff
@@ -0,0 +1,134 @@
+
+        val ipBlock = planner?.ipBlock
+        val ipBlockJson = JSONObject()
+        ipBlockJson.put("type", ipBlock?.type?.simpleName ?: "Error")
```
Consider using null instead of `Error`

Frederikam commented:
> Added RoutePlannerRestHandler (For monitoring RoutePlanner statistics / bans)

If you want to be able to monitor these things, you should consider adding Prometheus support

Frederikam commented:
Hey what happened to the fork and this PR?
